[MIG][11.0] Migrate stock_scanner

Garamotte opened the pull request:
Migration of the `stock_scanner` module to v11.0 (currently tested on master, as the 11.0 branch has not already been created on Odoo's repo).

Translation commits have been squashed.
The `scanner.scenario.custom` model has been removed as it wasn't used (as far as I know).

This PR contains changes added by https://github.com/OCA/stock-logistics-barcode/pull/108 on v10.0.

Commit:
[IMP] Add the user's homedir into available configuration file locations, in addition to the current working directory

Commit:
[IMP] Updated README file

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
Add 'stock_scanner/' from commit '607ce958c3d29d4202df677bc64009d9e4869d61'

git-subtree-dir: stock_scanner
git-subtree-mainline: ca2b48dde4fe5506256939a7bed714662fc042fa
git-subtree-split: 607ce958c3d29d4202df677bc64009d9e4869d61

Commit:
[MIG] Migrated stock_scanner to v9.0

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Rename manifest files

Commit:
[MIG] Migrate stock_scanner to v10.0

lasley committed:
[FIX] stock_scanner: Fix lint

gurneyalex committed:
stock_scanner sentinel: fix menu display

when the number of elements to display is long, the
'Selected' display is written on top of the last element.
This makes selecting the last element of the string difficult
because the label can be completely hidden from view.

This fixes the issue by reserving the last line for the Selection
display.

Commit:
handle the scrollbar properly

rousseldenis committed:
[10.0][FIX] stock_scanner Replace deprecated gtk icon

Commit:
[FIX] Fix v10.0 export scenario script

With the new API migration, some changes were not applied in the script
- The read method now always returns a list (even on a single id)
- The context argument is now at the end, so we need to add the load
    argument when not using named arguments

Commit:
[IMP] stock_scanner: Add unit tests

While writing the missing unit tests, some minor bugs were fixed.

Commit:
[MIG] stock_scanner: Migrate to v11.0

Garamotte commented:
The lint is red because of two external scripts that are not (yet) migrated to python3.
I'll migrate them using OdooRPC instead of openobject-library.

Commit:
[IMP] Add widget in python_code field for coding directly in view

benwillig committed:
[FIX] Handle KEY_RESIZE event in a selection action

benwillig committed:
[FIX] Avoid null chars in _input_text

Commit:
[IMP] Raise an error instead of simply displaying an error log line when some data is wrong in the scenario files

pedrobaeza committed:
Update __manifest__.py

Garamotte commented:
Updated to include last changes applied on v10.0 branch (editor widget for the python code fields, mute_logger in tests and null char fix).

The migration of sentinel.py to python3/OdooRPC is still to do.

Garamotte commented:
I fixed the `ir.cron` record: Cron jobs inherit from server actions in v11, so we have to replace the `function` and `args` fields by values like `state = code` and `code = model.function(args)`.

Garamotte commented:
The `export_scenario.py` and `sentinel.py` scripts are now migrated to Python3, using OdooRPC instead of openobject-library. I also fixed the licence blocks in some xml files.

@alnslang @lmignon @damdam-s This PR is now ready for reviews :)

Commit:
[IMP] stock_scanner: Add automatic json serialization for temporary values

This allows to access the temporary values by using 'json_tmp_valX',
     which will automatically (de)serialize the data. This is useful to
     store complex data structures in the temporary values, without
     having to manually do the serialization in the scenario's code.

lasley commented on stock_scanner/hardware/sentinel.py:
```diff
@@ -65,44 +53,26 @@ class Sentinel(object):
+            self.connection = odoorpc.ODOO.load(options.profile)
+        else:
+            raise Exception(
+                'Profile "{options.profile}" not found in file '
```
Should this be translated?

lasley commented on stock_scanner/models/scanner_hardware.py:
```diff
@@ -311,7 +305,6 @@ def _scanner_call(self, action, message=False,
         # Nothing matched, return an error
         return self._send_error(['Unknown', 'action'])
 
-    @api.multi
```
Any reason for removing the `api.multi`s? Explicit is better IMO

Garamotte replied:
I always remove `@api.multi` when migrating to v10.0 or higher because this is now the default.
Is there a guideline about this ?

lasley commented on stock_scanner/models/scanner_hardware.py:
```diff
@@ -469,11 +454,10 @@ def _do_scenario_save(self, message, transition_type, scenario_id=None,
-                         _('A001')])
+                    return self._send_error([
+                        _('No start step found'),
+                        _('on the scenario'),
```
Why are the strings being translated and sent through individually? (genuine question)

Garamotte replied:
This is because the very first version of the client didn't wrap and was made for hardware with small screens.
We can concatenate the strings now, thanks :)

lasley reviewed (approved):
Changes look good, thanks @sylvain-garancher. Few comments/questions inline, none blockers

lasley commented:
We also may need to consider #115

Garamotte commented:
@lasley Changes done.
I included my refactor of #115 for now, but I'll cherry-pick/rebase if it's modified again.

Garamotte commented:
Updated the patch from #115 and removed screen width/height fields from `scanner.hardware`.

Commit:
[IMP] Move the sentinel.py client to a separate python package named odoo-sentinel

Garamotte commented:
As the `sentinel.py` client is the same for all Odoo versions, and because the `openobject-library` python package is outdated and not longer installable (due to changes on PyPI), I extracted the python3/OdooRPC version from this PR and created a python package named `odoo-sentinel` and removed the `hardware` directory from the `stock_scanner` module.

The git repo is here: https://github.com/syleam/odoo-sentinel
The PyPI page is here: https://pypi.python.org/pypi/odoo-sentinel
It is only python3 compatible, and can be installed with a simple `pip install odoo-sentinel`.
Then, the client can be launched by running the `odoo-sentinel` command (with optional command line arguments).

I'll be happy to move the git repo and PyPI package ownership to the OCA, obviously :)

lasley commented:
I agree with the move. I think that it makes sense to host it on OCA side, seeing as it was already in OCA.

`OCA/odoo-sentinel`?

lmignon commented:
@sylvain-garancher The move is a good Idea and will ease a lot the installation of the sentinel client script. Thank you for this new package!

Garamotte commented:
I added the OCA as owner of the package on PyPI, but the transfer on GitHub must be initiated by someone who has the rights to create new repos on OCA.

lasley commented:
@sylvain-garancher - I can handle, but I'll need owner or admin rights for that repo I believe.

Garamotte commented:
@lasley I added you as admin on this repo, this should be sufficient (cf https://help.github.com/articles/transferring-a-repository-owned-by-your-organization/#transferring-to-another-organization).

Garamotte commented:
I'll re-add the width/height fields on `scanner.hardware` because we have an issue with Windows CE terminals: The TelnetCE application seems to announce a fixed size of 24x80 characters, and I didn't found a way to change that.

The behaviour will be:
- If the fields are defined, use this to force the window size
- If they are 0, automatically size the window

lasley commented:
@sylvain-garancher - I just went to do the transfer, and I don't have the `Settings` for that repo. Does your organization possibly fall under this subsection?

> If Allow members to delete or transfer repositories for this organization has been disabled in the originating organization, only organization owners can transfer organization repositories. If Allow members to create repositories for this organization has been disabled in the receiving organization, only organization owners can transfer repositories into the organization.

lasley commented:
Derp it seems I only half accepted the invitation. Don't mind me, nothing to see here!

benwillig committed:
[ADD] Temporary values as JSON

Garamotte commented:
For those who use this version of the module, I fixed a bug yesterday in the settings page, because the `set_values` method didn't call `super`. This prevented some settings from other modules to be applied.

This is the last issue we had, I think this PR would be merged now.

Garamotte commented:
Now that #127 is merged, I included it here.

@benwillig @lmignon I think I can remove the other `tmp_val*` fields now, what's your opinion about that ?

gurneyalex commented:
@sylvain-garancher if you do, please add a small migration note in the README and in the source code of scanner_hardware near the new serialized field to help ppl migrate their scenarios to the new version.

benwillig commented:
I'm agree in removing thoses fields too because they aren't explicit enough and it's hard to read the scenario with plenty of ```tmp_val*```.

elicoidal commented on stock_scanner/README.rst:
```diff
@@ -0,0 +1,244 @@
+^^^^^^
+
+Scenarios are automatically imported on a module update, like any other data.
+You just have to add the path to your `Scenario_Name.scenario` files in the `data` or `demo` sections in the `__openerp__.py` file.
```
Probably can be updated to `__manifest__.py`

gurneyalex committed:
update README.rst

change __openerp__.py to __manifest__.py

gurneyalex commented on stock_scanner/data/scenarios/Login/Login.scenario:
```diff
@@ -0,0 +1,19 @@
+  <Step id="stock_scanner.scanner_scenario_login_step_done" name="Authentication OK" no_back="False" step_back="False" step_start="False" step_stop="True"/>
+  <Step id="stock_scanner.scanner_scenario_login_step_login" name="Login name" no_back="True" step_back="True" step_start="True" step_stop="False"/>
+  <Step id="stock_scanner.scanner_scenario_login_step_pwd" name="Pwd" no_back="False" step_back="False" step_start="False" step_stop="False"/>
+  <Transition condition="terminal.check_credentials(terminal.tmp_val1, message)" from_id="stock_scanner.scanner_scenario_login_step_pwd" id="stock_scanner.scanner_scenario_transition_pwd_to_done" name="Authenticate" sequence="0" to_id="stock_scanner.scanner_scenario_login_step_done" tracer="" transition_type="keyboard"/>
```
needs updating if tmp_valx are removed

gurneyalex commented on stock_scanner/data/scenarios/Login/stock_scanner.scanner_scenario_login_step_done.py:
```diff
@@ -0,0 +1,13 @@
+'Put the returned result or message in <res>, as a list of strings.'
+'Put the returned value in <val>, as an integer'
+
+terminal.login(terminal.tmp_val1, message)
```
needs updating if tmp_valx are removed

gurneyalex commented on stock_scanner/data/scenarios/Login/stock_scanner.scanner_scenario_login_step_done.py:
```diff
@@ -0,0 +1,13 @@
+
+act = 'F'
+res = [
+    _('You are now authenticated as %s !') % terminal.tmp_val1,
```
needs updating if tmp_valx are removed

gurneyalex commented on stock_scanner/data/scenarios/Login/stock_scanner.scanner_scenario_login_step_pwd.py:
```diff
@@ -0,0 +1,14 @@
+'Put the returned result or message in <res>, as a list of strings.'
+'Put the returned value in <val>, as an integer'
+
+terminal.write({'tmp_val1': message})
```
needs updating if tmp_valx are removed

gurneyalex commented on stock_scanner/views/scanner_hardware.xml:
```diff
@@ -0,0 +1,101 @@
+                        <field name="tmp_val2"/>
+                        <field name="tmp_val3"/>
+                        <field name="tmp_val4"/>
+                        <field name="tmp_val5"/>
```
needs updating if tmp_valx are removed

gurneyalex committed:
[IMP] remove tmp_val[1-5]

* migrate Login / Logout scenario to use tmp_values
* rewrite the helper properties to store these values in the
  serialized fields tmp_values
* add a computed field to display the tmp_values in a table in
  the hardware form view.

gurneyalex commented:
I created https://github.com/syleam/stock-logistics-barcode/pull/1 for the removal of the tmp_valX fields

Garamotte commented:
Thanks @gurneyalex merged into the branch of this PR.

gurneyalex commented:
I think it's time this PR gets merged. I'll do so when the CI has finished running. :smile_cat:
